private/buf/buflsp/buflsp.go

```
@@ -0,0 +1,634 @@
+	s.folders = params.WorkspaceFolders
+	s.clientCap = params.Capabilities
+
+	// Always load the descriptor.proto file
+	if _, err := s.resolveImport(
+		ctx,
+		s.wellKnownTypesModuleSet,
+		s.wellKnownTypesBucket,
+		wellKnownTypesDescriptorProtoPath,
+	); err != nil {
+		return nil, err
+	}
+
+	// Reply with capabilities
+	initializeResult := &protocol.InitializeResult{
+		Capabilities: protocol.ServerCapabilities{
+			TextDocumentSync: &protocol.TextDocumentSyncOptions{
+				OpenClose: true,
+				Change:    protocol.TextDocumentSyncKindFull,
+			},
+			CompletionProvider: &protocol.CompletionOptions{
+				TriggerCharacters: []string{"."},
+			},
+			DocumentFormattingProvider: true,
+			DefinitionProvider:         true,
+			DocumentSymbolProvider:     true,
+			HoverProvider:              true,
+			SemanticTokensProvider:     &protocol.SemanticTokensOptions{},
+		},
+	}
+	return initializeResult, nil
+}
+
+func (s *server) Shutdown(ctx context.Context) error {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+	return s.fileWatcher.Close()
+}
+
+func (s *server) DidOpen(ctx context.Context, params *protocol.DidOpenTextDocumentParams) error {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+
+	// Check if it is already open.
+	if entry, ok := s.fileCache[params.TextDocument.URI.Filename()]; ok {
+		entry.refCount++
+		if _, err := entry.updateText(ctx, s, params.TextDocument.Text); err != nil {
+			return err
+		}
+		return nil
+	}
+	var moduleSet bufmodule.ModuleSet
+	wellKnownTypesCachePath := normalpath.Join(
+		s.container.CacheDirPath(),
+		lspWellKnownTypesCacheRelDirPath,
+	)
+	if strings.HasPrefix(normalpath.Normalize(params.TextDocument.URI.Filename()), wellKnownTypesCachePath) {
+		moduleSet = s.wellKnownTypesModuleSet
+	} else {
+		var err error
+		workspace, err := s.controller.GetWorkspace(ctx, params.TextDocument.URI.Filename())
+		if err != nil {
+			s.logger.Warn("could not determine workspace", zap.Error(err))
+			// Continue anyways if this fails.
```

Why?

---

I believe the reason for this is that we can still provide useful features even if we can't get a workspace.

---

The previous statement (that we still provide useful features even without a workspace) is still true, but this hack is no longer necessary as I've deferred workspace resolution until later, so we don't need to swallow this error here anymore.